clippy_utils/src/check_proc_macro.rs

```diff
@@ -250,6 +250,24 @@ fn variant_search_pat(v: &Variant<'_>) -> (Pat, Pat) {
     }
 }
 
+fn fn_kind_pat(tcx: TyCtxt<'_>, kind: &FnKind<'_>, body: &Body<'_>, hir_id: HirId) -> (Pat, Pat) {
+    let (start_pat, end_pat) = match kind {
+        FnKind::ItemFn(.., header) => (fn_header_search_pat(*header), Pat::Str("")),
+        FnKind::Method(.., sig) => (fn_header_search_pat(sig.header), Pat::Str("")),
+        FnKind::Closure => return (Pat::Str(""), expr_search_pat(tcx, &body.value).1),
+    };
+    let vis_span = match tcx.hir().get(hir_id) {
+        Node::Item(Item { vis_span, .. }) | Node::ImplItem(ImplItem { vis_span, .. }) => Some(vis_span),
+        Node::TraitItem(_) => None,
+        _ => unreachable!(),
```

Comment:
A few things:
* No need for an option here. The only thing being decided on here is the start pattern which can be assigned directly from the match.
* `TraitItem` should result in `start_pat`, not `pub`, as trait items can't have a visibility modifier.
* The default arm would be better as `Pat::Str("")` just in case something changes on the rustc side.

Reply:
Done!

> * The default arm would be better as `Pat::Str("")` just in case something changes on the rustc side.

Why `Pat::Str("")` instead of `start_pat`.

Reply:
It's an unexpected node which could have anything at the start of it. Since there's nothing which will definitely match if it's not a proc-macro `Pat::Str("")` is used to match anything.

Reply:
I see! Thanks!